scipy/_lib/_array_api.py

```diff
@@ -270,6 +281,8 @@ def xp_assert_equal(actual, desired, *, check_namespace=True, check_dtype=True,
     __tracebackhide__ = True  # Hide traceback for py.test
     if xp is None:
         xp = array_namespace(actual)
+    if isinstance(desired, list):
+        desired = xp.asarray(desired, dtype=xp.asarray(actual).dtype)
```

Comment:
I'm not convinced we should do this without adding separate namespace checks in the tests. And it seems cleaner to me to keep them together.

Reply:
Not having this IMO inflates the diffs and creates a ton of busywork. Not sure what separate namespace checks mean exactly either, care to send a patch?

Reply:
To demonstrate the problem, let's look at this test and treat `windows.barthann` as a black box:

```python
def test_basic(self, xp):
    xp_assert_close(windows.barthann(6, sym=True, xp=xp),
                    [0, 0.35857354213752, 0.8794264578624801,
                     0.8794264578624801, 0.3585735421375199, 0],
                    rtol=1e-15, atol=1e-15)
```

Suppose `windows.barthann(6, sym=True, xp=torch)` returns a NumPy array. The test currently does not fail since the automatic coercion of lists you've added will just result in checking NumPy against NumPy.

> Not sure what separate namespace checks mean

you could, after each `xp_assert_close` call, also do

```python
assert array_namespace(actual) == array_namespace(xp.empty(0))
```

but I don't prefer that.

> creates a ton of busywork

Unfortunately so - let me know if you come up with any ideas which avoid it!

Reply:
Okay, agreed on this being a reason indeed. The error is frequent, and it's something we do want to catch in tests. 

I played a bit with adding a namespace check to `xp_assert_close`: it accepts `xp=None` kwarg, so in principle one can reuse it: 

```
if isinstance(desired, list):
     _assert_matching_namespace(actual, xp.empty(0))
    desired = xp.asarray(desired, dtype=xp.asarray(actual).dtype)
```

and then tests will need to be written with an explicit `xp`:
```
xp_assert_close(xp.arange(3), [0, 1, 2], xp=xp)
```

However this looks brittle, error prone and a non-idiomatic reuse of the `xp` optional argument. 

So --- however grudgingly --- I agree, it's best to just add missing `xp.asarray` calls. Which the tip of this PR does now.